### PR TITLE
[perlin] Add industry noise and circular gradient

### PR DESCRIPTION
**Pull request by falkecarlsen**
- [x] Add distinct bases for population and industry to emulate seeding of Perlin noise
- [x] Add industry noise application
- [x] Add circular gradient centered on network coordinate from parameter
- [ ] Unit-test

### CODE REVIEW

**Comment by falkecarlsen** [comment]
#8 is relevant for this PR

File: perlin.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+# The 'noise' lib has good resolution until above 10 mil, but a SIGSEGV is had on values above [-100000, 100000]
+# FIXME: these values result in a noisemap which contains CRT-like lines and patterns. Find better, sane values.
+# POPULATION_BASE = random.randrange(-1000, 1000)
+POPULATION_BASE: float = 0
+# INDUSTRY_BASE = random.randrange(-1000, 1000)
+INDUSTRY_BASE: float = 0
+
+
+def get_edge_pair_centroid(coords: list) -> (float, float):
```
Comment (NicEastvillage) [inline perlin.py]: `list` is a function. The type hint for lists are `List[T]` and for tuples are `Tuple[T, U, ..]`. so it should be:
```python
def get_edge_pair_centroid(coords: List[Tuple[float, float]]) -> Tuple[float, float]:
    ...
```
This is probably relevant other places as well, but I don't want to be super strict about it. Just wanted to let you know

File: perlin.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    :return: the radius of the network
+    """
+    return np.max([distance(centre, node.getCoord()) for node in net.getNodes()])
+
+
+def distance(pos1, pos2):
+    """
+    Get hypotenuse using Phytagoras theorem
+    :return: the distance between pos1 and pos2
+    """
```
Comment (NicEastvillage) [inline perlin.py]: > Get hypotenuse using Phytagoras theorem

Eh. It returns the *length* of the hypotenuse which in this case is equal to the euclidean distance between pos1 and pos2. So why no just say it returns the distance?

File: perlin.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    x1, y1 = pos1
+    x2, y2 = pos2
+    return np.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)
+
+
+def display_noisemap(net: sumolib.net.Net, scale: float = 0.005, octave: int = 3):
```
Comment (NicEastvillage) [inline perlin.py:126]: This function renders a massive image. It takes 99% of the running time and even crashed the program one time when I ran it on some randomly generated networks. Let's not call it until we make it a bit more sane.

File: randomActivityGen.py [hunk 4]
```diff
@@ -106,8 +106,14 @@ def setup_city_gates(net: sumolib.net.Net, stats: ET.ElementTree, gate_count: in
...
-    apply_perlin_noise(net, stats)
+    # Scale and octave seems like sane values for the moment
+    apply_network_noise(net, stats, 0.005, 3)
+
     setup_city_gates(net, stats, int(args["--gates.count"]))
 
     # Write statistics back
     stats.write(args["--output-file"])
+
+    centre = (500, 500)
```
Comment (NicEastvillage) [inline randomActivityGen.py]: This center seem to have no purpose

File: render.py [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+                elif world[i][j] < threshold + 1.0:
+                    color_world[i][j] = snow
+
+        return color_world
+
+    island_world_grad = add_color2(world_noise)
+    toimage(island_world_grad).show()
+
+
+def trash():
```
Comment (NicEastvillage) [inline render.py:335]: What is this function for?

**Review by NicEastvillage — CHANGES_REQUESTED** [review]
Hey. It looks good. I put a few comments but it is mostly about small things. Then only thing I want you to change is removing the call to `display_noisemap` because that function is not ideal.

File: perlin.py (hunk 3 above)
Comment (falkecarlsen) [reply perlin.py:126]: The high runtime makes sense, as it samples many more edges than the network contains. I agree that it shouldn't be run by default, though.

File: randomActivityGen.py (hunk 4 above)
Comment (falkecarlsen) [reply randomActivityGen.py]: Debug leftovers

File: render.py (hunk 5 above)
Comment (falkecarlsen) [reply render.py:335]: Debug leftovers

File: perlin.py (hunk 2 above)
Comment (falkecarlsen) [reply perlin.py]: Fixed in d8433396bbd5e96b6ca82bf3749a4a26fee92ab3

File: randomActivityGen.py (hunk 4 above)
Comment (falkecarlsen) [reply randomActivityGen.py]: Fixed in 1156f14

File: perlin.py (hunk 3 above)
Comment (falkecarlsen) [reply perlin.py:126]: Fixed in 1156f14

File: perlin.py (hunk 1 above)
Comment (falkecarlsen) [reply perlin.py]: Cool, I tried fiddling around with it but didn't realise that `list != List`.
Fixed in 1ceff2d, it seems that's the only place that type is passed.

**Comment by falkecarlsen** [comment]
Note also that bases are both set to 0 right now. We should fix #8 soon.

**Review by NicEastvillage — APPROVED** [review]

**Pull request merged**